common/attest_plugin.c

```diff
@@ -0,0 +1,441 @@
+        OE_RAISE(OE_INVALID_PARAMETER);
+
+    // Find the guid and remove it.
+    cur = _find_plugin(*list, &plugin->format_id, &prev);
+    if (cur == NULL)
+        OE_RAISE(OE_NOT_FOUND);
+
+    if (prev != NULL)
+        prev->next = cur->next;
+    else
+        *list = cur->next;
+
+    // Run the unregister hook for the plugin.
+    OE_CHECK(cur->plugin->on_unregister(cur->plugin));
+
+    result = OE_OK;
+
+done:
+    oe_free(cur);
+    return result;
+}
+
+oe_result_t oe_register_attester(
+    oe_attester_t* plugin,
+    const void* config_data,
+    size_t config_data_size)
+{
+    return _register_plugin(
+        &attesters,
+        (oe_attestation_role_t*)plugin,
+        config_data,
+        config_data_size);
+}
+
+oe_result_t oe_register_verifier(
+    oe_verifier_t* plugin,
+    const void* config_data,
+    size_t config_data_size)
+{
+    return _register_plugin(
+        &verifiers,
+        (oe_attestation_role_t*)plugin,
+        config_data,
+        config_data_size);
+}
+
+oe_result_t oe_unregister_attester(oe_attester_t* plugin)
+{
+    return _unregister_plugin(&attesters, (oe_attestation_role_t*)plugin);
+}
+
+oe_result_t oe_unregister_verifier(oe_verifier_t* plugin)
+{
+    return _unregister_plugin(&verifiers, (oe_attestation_role_t*)plugin);
+}
+
+static oe_result_t _wrap_with_header(
+    const uuid_t* format_id,
+    const uint8_t* data,
+    size_t data_size,
+    uint8_t** total_data,
+    size_t* total_data_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    oe_attestation_header_t* header;
+
+    OE_CHECK(oe_safe_add_sizet(sizeof(*header), data_size, total_data_size));
+
+    *total_data = (uint8_t*)oe_malloc(*total_data_size);
+    if (*total_data == NULL)
+        OE_RAISE(OE_OUT_OF_MEMORY);
+
+    header = (oe_attestation_header_t*)*total_data;
+    header->version = OE_ATTESTATION_HEADER_VERSION;
+    header->format_id = *format_id;
+    header->data_size = data_size;
+    memcpy(header->data, data, data_size);
+
+    result = OE_OK;
+
+done:
+    return result;
+}
+
+oe_result_t oe_get_evidence(
+    const uuid_t* format_id,
+    uint32_t flags,
+    const oe_claim_t* custom_claims,
+    size_t custom_claims_length,
+    const void* opt_params,
+    size_t opt_params_size,
+    uint8_t** evidence_buffer,
+    size_t* evidence_buffer_size,
+    uint8_t** endorsements_buffer,
+    size_t* endorsements_buffer_size)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    struct plugin_list_node_t* plugin_node = NULL;
+    oe_attester_t* plugin = NULL;
+    uint8_t* plugin_evidence = NULL;
+    size_t plugin_evidence_size = 0;
+    uint8_t* plugin_endorsements = NULL;
+    size_t plugin_endorsements_size = 0;
+    uint8_t* total_evidence_buf = NULL;
+    size_t total_evidence_size = 0;
+    uint8_t* total_endorsements_buf = NULL;
+    size_t total_endorsements_size = 0;
+
+    if (!format_id || !evidence_buffer || !evidence_buffer_size ||
+        (endorsements_buffer && !endorsements_buffer_size))
```

Comment:
>endorsements_buffer && !endorsements_buffer_size [](start = 9, length = 48)

Consider having them both valid or both invalid.

Reply:
I added a check for seeing if both are valid or both are invalid.